File: code/matlab/problem2_2.m
% Beamforming

% clear
clear all;
clc;
close all;

% setup
NUM_ANA= 40; % antenna number
CARR_FREQ= 2.4e9; % Hz. carrier frequency
WAVE_LEN= physconst("LightSpeed")/CARR_FREQ; % m. wavelength
ANTENNA_DIS= WAVE_LEN/2; % distance between two antennas



% the beamwidth is from theta_l to theta_u degrees
theta_l=0;
theta_u=40;

% Angle at which the beam position is maximum
ANGLE_DES= (theta_l + theta_u)/2; % at beam center 


% Steering vector along the direction of desired signal
Steering_des=  [exp(-j*[0:NUM_ANA-1].'*2*pi*ANTENNA_DIS*sin(ANGLE_DES*pi/180)/WAVE_LEN)];


% Final Beam Steering Vector (or Array Response Vector)
Beam_weight_MF= Steering_des/NUM_ANA;

%Problem No. 2
% Minimize average sidelobe 
% accumulate all the phases for the sidelobe (interferers)
P_matrix= zeros(NUM_ANA, NUM_ANA);

% interferer from -90 to theta_l degrees
for i=-90:1:theta_l
    P_matrix= P_matrix + ...
        exp(-j*[0:NUM_ANA-1]'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN) *  ...
        exp(-j*[0:NUM_ANA-1]'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)';
end


for i= theta_u:1:90
    P_matrix= P_matrix + ...
        exp(-j*[0:NUM_ANA-1]'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN) * ... 
        exp(-j*[0:NUM_ANA-1]'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)';
end
% optimal problem
cvx_begin
    variable w(NUM_ANA,1) complex
    expression Match_output;
    Match_output=quad_form(w,P_matrix);
    minimize(Match_output);
    subject to
    w'*Steering_des==1;
cvx_end

% plot angle spectrum
steering_vec_plot=[];
for i=-90:1:90
    steering_vec_plot=[steering_vec_plot; exp(-j*[0:NUM_ANA-1]*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)];
end




% plot
figure(1)

plot([-90:1:90],10*log10(abs(w'*steering_vec_plot.').^2));
title('Minimizing the Average Sidelobe')

fmt = 'M=40, \theta_{d}=20, \theta_{1}=1, \theta_{u}=30';
xlabel('Angle(degree)');
ylabel('Angle Response(db)');

hold on
grid on
xlim([-90 90])
ylim([-120, 0])

% save the figure
saveas(gcf, 'problem2_2.png')
savefig(gcf, 'problem2_2.fig')

File: code/python/problem2_2.py
"""Average sidelobe suppression for a wide (0°–40°) mainlobe."""

import pickle

import numpy as np
import cvxpy as cp
import matplotlib.pyplot as plt


NUM_ANA = 40
CARR_FREQ = 2.4e9  # Hz
LIGHT_SPEED = 299_792_458.0  # m/s
WAVE_LEN = LIGHT_SPEED / CARR_FREQ
ANTENNA_DIS = WAVE_LEN / 2

theta_l = 0
theta_u = 40
ANGLE_DES = (theta_l + theta_u) / 2

indices = np.arange(NUM_ANA)
# Steering_des implements a(θ) for the uniform linear array at ANGLE_DES.
steering_phase = -1j * 2 * np.pi * ANTENNA_DIS * np.sin(np.deg2rad(ANGLE_DES)) / WAVE_LEN
Steering_des = np.exp(steering_phase * indices)
Beam_weight_MF = Steering_des / NUM_ANA


def steering_column(angle_deg: float) -> np.ndarray:
    """Column steering vector a(θ) = exp(-j2πd sinθ / λ · antenna_index)."""
    phase = -1j * 2 * np.pi * ANTENNA_DIS * np.sin(np.deg2rad(angle_deg)) / WAVE_LEN
    return np.exp(phase * indices)[:, None]


# Collect sidelobe energy contributions from outside the 0°–40° sector.
P_matrix = np.zeros((NUM_ANA, NUM_ANA), dtype=np.complex128)
for angle in range(-90, theta_l + 1):
    vec = steering_column(angle)
    P_matrix += vec @ vec.conj().T
for angle in range(theta_u, 91):
    vec = steering_column(angle)
    P_matrix += vec @ vec.conj().T

w = cp.Variable(NUM_ANA, complex=True)
objective = cp.Minimize(cp.quad_form(w, P_matrix))
# Enforce unity response toward the beam center to protect the desired signal.
constraints = [cp.sum(cp.multiply(np.conj(Steering_des), w)) == 1]
problem = cp.Problem(objective, constraints)
problem.solve(solver=cp.CLARABEL)

# Plot the full beampattern to see how widening the mainlobe affects sidelobes.
angles = np.arange(-90, 91)
angles_rad = np.deg2rad(angles)
steering_vec_plot = np.exp(
    -1j * np.outer(np.sin(angles_rad), indices) * 2 * np.pi * ANTENNA_DIS / WAVE_LEN
)
pattern_db = 10 * np.log10(np.abs(np.conj(w.value) @ steering_vec_plot.T) ** 2 + 1e-12)

fig, ax = plt.subplots()
ax.plot(angles, pattern_db)
ax.set_title("Minimizing the Average Sidelobe")
fmt = (
    rf"$M={NUM_ANA},\ \theta_d={ANGLE_DES},\ "
    rf"\theta_\ell={theta_l},\ \theta_u={theta_u}$"
)
ax.set_xlabel("Angle (degree)")
ax.set_ylabel("Angle Response (dB)")
ax.text(-80, -10, fmt)
ax.grid(True)
ax.set_xlim([-90, 90])
ax.set_ylim([-120, 0])
fig.savefig("problem2_2.png", dpi=300, bbox_inches="tight")
with open("problem2_2.fig.pickle", "wb") as fig_file:
    pickle.dump(fig, fig_file)
plt.show()
plt.close(fig)
